pkg/session/session.go

```diff
@@ -95,28 +127,53 @@ func GetOrCreate(
 	return &session, nil
 }
 
-func newClient(ctx context.Context, url *url.URL, thumprint string) (*govmomi.Client, error) {
+func newClient(ctx GetOrCreateContext, sessionKey string, url *url.URL, thumprint string, feature Feature) (*govmomi.Client, error) {
 	insecure := thumprint == ""
 	soapClient := soap.NewClient(url, insecure)
 	if !insecure {
 		soapClient.SetThumbprint(url.Host, thumprint)
 	}
 
-	vimClient, err := vim25.NewClient(ctx, soapClient)
+	vimClient, err := vim25.NewClient(ctx.context, soapClient)
 	if err != nil {
 		return nil, err
 	}
+
 	c := &govmomi.Client{
 		Client:         vimClient,
 		SessionManager: session.NewManager(vimClient),
 	}
-	if err := c.Login(ctx, url.User); err != nil {
+
+	if feature.EnableKeepAlive {
+		vimClient.RoundTripper = session.KeepAliveHandler(vimClient.RoundTripper, feature.KeepAliveDuration, func(tripper soap.RoundTripper) error {
+			// we tried implementing
+			// c.Login here but the client once logged out
+			// keeps errong in invalid username or password
+			// we tried with cached username and password in session still the error persisted
+			// hence we just clear the cache and expect the client to
+			// be recreated in next GetOrCreate call
+			_, err := methods.GetCurrentTime(ctx.context, tripper)
+			if err != nil {
+				ctx.logger.Error(err, "failed to keep alive govmomi client")
+				ClearCache(sessionKey)
+			}
+			return err
+		})
+	}
+
+	if err := c.Login(ctx.context, url.User); err != nil {
 		return nil, err
 	}
 
 	return c, nil
 }
 
+func ClearCache(sessionKey string) {
```

Comment:
```suggestion
func clearCache(sessionKey string) {
```

Reply:
Ack

Reply:
Disregard this one, we do not wanna expose the clearCache method outside the package.